From 7d3f53cb9add44791ef755684c4f1539f77ef0ee Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 16 Jun 2020 11:04:16 +0200
Subject: [PATCH 022/217] rcu: irq_pipeline: handle non-maskable pipeline entry
 context

From the standpoint of the in-band context, an interrupt entry is
logically similar to a NMI when pipelining is enabled, since it may
preempt the in-band stage almost anywhere as IRQs are only virtually
masked most of the time there, including inside (virtually)
interrupt-free sections.

Enable the RCU code to detect a pipeline entry, handling it like a
NMI, so that we may take the RCU read lock there as well.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>

rcu: irq_pipeline: enable read lock sections from out-of-band context

Allow an out-of-band caller to start a RCU read locked section by
forcing a RCU entry from NMI. Conversely, leaving such section unwinds
the NMI entry. Like RCU read lock sections, NMI entries can be nested.

In the same move, stop assuming that any out-of-band context means
that RCU is not watching anymore, as a matter of fact the previous
change would contradict this.

The logic is based on the invariant that the pipeline stage cannot
change while in a RCU read locked section.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/linux/hardirq.h   | 10 ++++++++++
 include/linux/rcupdate.h  | 14 +++++++++++++-
 kernel/context_tracking.c |  6 +++---
 kernel/rcu/tree.c         |  7 +++++--
 kernel/rcu/tree_plugin.h  |  7 ++++---
 kernel/rcu/update.c       | 31 +++++++++++++++++++++++++++++++
 6 files changed, 66 insertions(+), 9 deletions(-)

diff --git a/include/linux/hardirq.h b/include/linux/hardirq.h
index c6c4e9143..2ba79d080 100644
--- a/include/linux/hardirq.h
+++ b/include/linux/hardirq.h
@@ -148,4 +148,14 @@ static inline bool on_pipeline_entry(void)
 	return irqs_pipelined() && in_pipeline();
 }
 
+static inline bool in_nonmaskable(void)
+{
+	/*
+	 * From the standpoint of tne in-band stage, any oob activity
+	 * stands for a non-maskable event being handled (including
+	 * the pipeline entry context).
+	 */
+	return running_oob() || in_nmi();
+}
+
 #endif /* LINUX_HARDIRQ_H */
diff --git a/include/linux/rcupdate.h b/include/linux/rcupdate.h
index 0844ab328..9e32a025a 100644
--- a/include/linux/rcupdate.h
+++ b/include/linux/rcupdate.h
@@ -160,6 +160,14 @@ static inline void rcu_nocb_flush_deferred_wakeup(void) { }
 
 #endif /* #else #ifdef CONFIG_RCU_NOCB_CPU */
 
+#ifdef CONFIG_IRQ_PIPELINE
+void rcu_oob_prepare_lock(void);
+void rcu_oob_finish_lock(void);
+#else
+#define rcu_oob_prepare_lock()	 do { } while (0)
+#define rcu_oob_finish_lock()	 do { } while (0)
+#endif
+
 /*
  * Note a quasi-voluntary context switch for RCU-tasks's benefit.
  * This is a macro rather than an inline function to avoid #include hell.
@@ -371,7 +379,7 @@ static inline int rcu_read_lock_bh_held(void)
 
 static inline int rcu_read_lock_sched_held(void)
 {
-	return !preemptible();
+	return !running_inband() || !preemptible();
 }
 
 static inline int rcu_read_lock_any_held(void)
@@ -844,6 +852,7 @@ do {									      \
  */
 static __always_inline void rcu_read_lock(void)
 {
+	rcu_oob_prepare_lock();
 	__rcu_read_lock();
 	__acquire(RCU);
 	rcu_lock_acquire(&rcu_lock_map);
@@ -880,6 +889,7 @@ static inline void rcu_read_unlock(void)
 	rcu_lock_release(&rcu_lock_map); /* Keep acq info for rls diags. */
 	__release(RCU);
 	__rcu_read_unlock();
+	rcu_oob_finish_lock();
 }
 
 /**
@@ -937,6 +947,7 @@ static inline void rcu_read_unlock_bh(void)
 static inline void rcu_read_lock_sched(void)
 {
 	preempt_disable();
+	rcu_oob_prepare_lock();
 	__acquire(RCU_SCHED);
 	rcu_lock_acquire(&rcu_sched_lock_map);
 	RCU_LOCKDEP_WARN(!rcu_is_watching(),
@@ -961,6 +972,7 @@ static inline void rcu_read_unlock_sched(void)
 			 "rcu_read_unlock_sched() used illegally while idle");
 	rcu_lock_release(&rcu_sched_lock_map);
 	__release(RCU_SCHED);
+	rcu_oob_finish_lock();
 	preempt_enable();
 }
 
diff --git a/kernel/context_tracking.c b/kernel/context_tracking.c
index 938c48952..5139dd2b0 100644
--- a/kernel/context_tracking.c
+++ b/kernel/context_tracking.c
@@ -239,7 +239,7 @@ void noinstr ct_nmi_exit(void)
 	ct_kernel_exit_state(CT_RCU_WATCHING);
 	// ... but is no longer watching here.
 
-	if (!in_nmi())
+	if (!in_nonmaskable())
 		rcu_task_exit();
 }
 
@@ -273,7 +273,7 @@ void noinstr ct_nmi_enter(void)
 	 */
 	if (!rcu_is_watching_curr_cpu()) {
 
-		if (!in_nmi())
+		if (!in_nonmaskable())
 			rcu_task_enter();
 
 		// RCU is not watching here ...
@@ -287,7 +287,7 @@ void noinstr ct_nmi_enter(void)
 		instrument_atomic_write(&ct->state, sizeof(ct->state));
 
 		incby = 1;
-	} else if (!in_nmi()) {
+	} else if (!in_nonmaskable()) {
 		instrumentation_begin();
 		rcu_irq_enter_check_tick();
 	} else  {
diff --git a/kernel/rcu/tree.c b/kernel/rcu/tree.c
index 7b073b8b5..57540186a 100644
--- a/kernel/rcu/tree.c
+++ b/kernel/rcu/tree.c
@@ -649,7 +649,7 @@ void __rcu_irq_enter_check_tick(void)
 	struct rcu_data *rdp = this_cpu_ptr(&rcu_data);
 
 	// If we're here from NMI there's nothing to do.
-	if (in_nmi())
+	if (in_nonmaskable())
 		return;
 
 	RCU_LOCKDEP_WARN(!rcu_is_watching_curr_cpu(),
@@ -733,6 +733,9 @@ notrace bool rcu_is_watching(void)
 {
 	bool ret;
 
+	if (on_pipeline_entry())
+ 		return true;
+ 
 	preempt_disable_notrace();
 	ret = rcu_is_watching_curr_cpu();
 	preempt_enable_notrace();
@@ -4795,7 +4798,7 @@ bool rcu_lockdep_current_cpu_online(void)
 	struct rcu_data *rdp;
 	bool ret = false;
 
-	if (in_nmi() || !rcu_scheduler_fully_active)
+	if (in_nonmaskable() || !rcu_scheduler_fully_active)
 		return true;
 	preempt_disable_notrace();
 	rdp = this_cpu_ptr(&rcu_data);
diff --git a/kernel/rcu/tree_plugin.h b/kernel/rcu/tree_plugin.h
index 2d865b209..d338c2d5d 100644
--- a/kernel/rcu/tree_plugin.h
+++ b/kernel/rcu/tree_plugin.h
@@ -663,8 +663,9 @@ static void rcu_read_unlock_special(struct task_struct *t)
 	bool preempt_bh_were_disabled =
 			!!(preempt_count() & (PREEMPT_MASK | SOFTIRQ_MASK));
 
-	/* NMI handlers cannot block and cannot safely manipulate state. */
-	if (in_nmi())
+	/* Neither NMI handlers nor pipeline entry code can either
+	   block or safely manipulate state. */
+	if (in_nonmaskable())
 		return;
 
 	local_irq_save(flags);
@@ -852,7 +853,7 @@ void rcu_read_unlock_strict(void)
 {
 	struct rcu_data *rdp;
 
-	if (irqs_disabled() || in_atomic_preempt_off() || !rcu_state.gp_kthread)
+	if (on_pipeline_entry() || running_oob() || irqs_disabled() || in_atomic_preempt_off() || !rcu_state.gp_kthread)
 		return;
 
 	/*
diff --git a/kernel/rcu/update.c b/kernel/rcu/update.c
index f8436969e..5cef5e6af 100644
--- a/kernel/rcu/update.c
+++ b/kernel/rcu/update.c
@@ -102,6 +102,11 @@ module_param(rcu_normal_after_boot, int, 0444);
  */
 static bool rcu_read_lock_held_common(bool *ret)
 {
+	if (irqs_pipelined() &&
+		(hard_irqs_disabled() || running_oob())) {
+		*ret = true;
+		return true;
+	}
 	if (!debug_lockdep_rcu_enabled()) {
 		*ret = true;
 		return true;
@@ -250,6 +255,32 @@ EXPORT_SYMBOL_GPL(rcu_inkernel_boot_has_ended);
 
 #endif /* #ifndef CONFIG_TINY_RCU */
 
+#ifdef CONFIG_IRQ_PIPELINE
+
+/*
+ * Prepare for taking the RCU read lock when running out-of-band. Nop
+ * otherwise.
+ */
+void rcu_oob_prepare_lock(void)
+{
+	if (!on_pipeline_entry() && running_oob())
+		ct_nmi_enter();
+}
+EXPORT_SYMBOL_GPL(rcu_oob_prepare_lock);
+
+/*
+ * Converse to rcu_oob_prepare_lock(), after dropping the RCU read
+ * lock.
+ */
+void rcu_oob_finish_lock(void)
+{
+	if (!on_pipeline_entry() && running_oob())
+		ct_nmi_exit();
+}
+EXPORT_SYMBOL_GPL(rcu_oob_finish_lock);
+
+#endif	/* CONFIG_IRQ_PIPELINE */
+
 /*
  * Test each non-SRCU synchronous grace-period wait API.  This is
  * useful just after a change in mode for these primitives, and
-- 
2.30.2

